integration_test/tests/disconnect_tests.rs

```diff
@@ -0,0 +1,130 @@
+        }
+    }
+
+    #[test]
+    fn sql_disconnect_implicitly_closing_statment_cursors_is_safe_for_closed_statements() {
+        let (env, dbc, stmt) = default_setup_connect_and_alloc_stmt(AttrOdbcVersion::SQL_OV_ODBC3);
+        let query = b"SELECT * FROM integration_test.foo\0".map(|b| b as u16);
+        unsafe {
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLExecDirectW(stmt, query.as_ptr(), SQL_NTS as i32,),
+                "{}",
+                get_sql_diagnostics(HandleType::SQL_HANDLE_STMT, *stmt.cast::<Handle>()),
+            );
+
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLFreeStmt(stmt, FreeStmtOption::SQL_CLOSE as i16),
+                "{}",
+                get_sql_diagnostics(HandleType::SQL_HANDLE_STMT, *stmt.cast::<Handle>()),
+            );
+
+            // we will not free the statement, instead proceeding directly to disconnect and freeing the dbc and env handles
+            disconnect_and_free_dbc_and_env_handles(env, dbc)
+        }
+    }
+
+    #[test]
+    fn sql_disconnect_implicitly_closing_statment_cursors_is_safe_for_unbound_statemens() {
+        let (env, dbc, stmt) = default_setup_connect_and_alloc_stmt(AttrOdbcVersion::SQL_OV_ODBC3);
+        let query = b"SELECT * FROM integration_test.foo\0".map(|b| b as u16);
+        unsafe {
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLExecDirectW(stmt, query.as_ptr(), SQL_NTS as i32,),
+                "{}",
+                get_sql_diagnostics(HandleType::SQL_HANDLE_STMT, *stmt.cast::<Handle>()),
+            );
+
+            // we didn't actually bind a statement, so this should be a no-op
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLFreeStmt(stmt, FreeStmtOption::SQL_UNBIND as i16),
+                "{}",
+                get_sql_diagnostics(HandleType::SQL_HANDLE_STMT, *stmt.cast::<Handle>()),
+            );
+
+            // we will not free the statement, instead proceeding directly to disconnect and freeing the dbc and env handles
+            disconnect_and_free_dbc_and_env_handles(env, dbc)
+        }
+    }
+
+    #[test]
+    fn sql_disconnect_handles_many_statements_properly() {
+        let env = allocate_env(AttrOdbcVersion::SQL_OV_ODBC3);
+        let dbc = connect_with_conn_string(env, None)
+            .expect("Failed to connect with default connection string");
+        let stmt_1 = allocate_statement(dbc).expect("Failed to allocate statement 1");
+        let stmt_2 = allocate_statement(dbc).expect("Failed to allocate statement 2");
+        let stmt_3 = allocate_statement(dbc).expect("Failed to allocate statement 3");
+        let stmt_4 = allocate_statement(dbc).expect("Failed to allocate statement 4");
+        let stmt_5 = allocate_statement(dbc).expect("Failed to allocate statement 5");
+        let statements = [stmt_1, stmt_2, stmt_3, stmt_4, stmt_5];
```

Comment:
```suggestion
        let stmt_4 = allocate_statement(dbc).expect("Failed to allocate statement 4");
        let statements = [stmt_1, stmt_2, stmt_3, stmt_4];
```
It seems `stmt_5` are extraneous. With just 4 we achieve everything you're trying to test:
- multiple statements are opened and have queries executed on them
- one is closed via SQLFreeStmt
- one is closed via SQLFreeStmt and then free via SQLFreeHandle
- multiple are closed implicitly via SQLDisconnect

Is having a 5th statement absolutely necessary to test some behavior I'm missing?

Reply:
It's just a statement that doesn't have anything done with it.